File: dataOds/InLong-versions/apache-inlong-1.2.0-incubating/inlong-tubemq/tubemq-client-twins/tubemq-client-go/metadata/partition.go
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package metadata

import (
	"math"
	"strconv"
	"strings"
	"time"

	"github.com/apache/incubator-inlong/inlong-tubemq/tubemq-client-twins/tubemq-client-go/errs"
	"github.com/apache/incubator-inlong/inlong-tubemq/tubemq-client-twins/tubemq-client-go/flowctrl"
	"github.com/apache/incubator-inlong/inlong-tubemq/tubemq-client-twins/tubemq-client-go/util"
)

// Partition represents the metadata of a partition.
type Partition struct {
	topic        string
	broker       *Node
	partitionID  int32
	partitionKey string
	offset       int64
	lastConsumed bool
	consumeData  *ConsumeData
	flowCtrl     *flowctrl.Result
	freqCtrl     *flowctrl.Item
	strategyData *strategyData
	totalZeroCnt int32
}

// ConsumeData represents the consumption metadata of a partition.
type ConsumeData struct {
	time        int64
	errCode     int32
	escLimit    bool
	msgSize     int32
	dltLimit    int64
	curDataDlt  int64
	requireSlow bool
}

type strategyData struct {
	nextStageUpdate   int64
	nextSliceUpdate   int64
	limitSliceMsgSize int64
	curStageMsgSize   int64
	curSliceMsgSize   int64
}

// NewConsumeData returns a consume data.
func NewConsumeData(time int64, errCode int32, escLimit bool, msgSize int32, dltLimit int64,
	curDataDlt int64, requireSlow bool) *ConsumeData {
	return &ConsumeData{
		time:        time,
		errCode:     errCode,
		escLimit:    escLimit,
		msgSize:     msgSize,
		dltLimit:    dltLimit,
		curDataDlt:  curDataDlt,
		requireSlow: requireSlow,
	}
}

// NewPartition parses a partition from the given string.
// The format of partition string: brokerInfo#topic:partitionId
func NewPartition(partition string) (*Partition, error) {
	s := strings.Split(partition, "#")
	if len(s) == 1 {
		return nil, errs.ErrInvalidPartitionString
	}
	b, err := NewNode(true, s[0])
	if err != nil {
		return nil, err
	}
	topicPartitionID := strings.Split(s[1], ":")
	if len(topicPartitionID) == 1 {
		return nil, errs.ErrInvalidPartitionString
	}
	topic := topicPartitionID[0]
	partitionID, err := strconv.Atoi(topicPartitionID[1])
	if err != nil {
		return nil, err
	}
	partitionKey := strconv.Itoa(int(b.id)) + ":" + topic + ":" + strconv.Itoa(partitionID)
	item := flowctrl.NewItem()
	item.SetTp(flowctrl.RequestFrequencyControl)
	return &Partition{
		topic:        topic,
		broker:       b,
		partitionID:  int32(partitionID),
		strategyData: &strategyData{},
		partitionKey: partitionKey,
		consumeData: &ConsumeData{
			curDataDlt: util.InvalidValue,
		},
		freqCtrl: item,
	}, nil
}

// GetLastConsumed returns lastConsumed of a partition.
func (p *Partition) GetLastConsumed() bool {
	return p.lastConsumed
}

// GetPartitionID returns the partition id of a partition.
func (p *Partition) GetPartitionID() int32 {
	return p.partitionID
}

// GetPartitionKey returns the partition key of a partition.
func (p *Partition) GetPartitionKey() string {
	return p.partitionKey
}

// GetTopic returns the topic of the partition subscribed to.
func (p *Partition) GetTopic() string {
	return p.topic
}

// GetBroker returns the broker.
func (p *Partition) GetBroker() *Node {
	return p.broker
}

// String returns the metadata of a Partition as a string.
func (p *Partition) String() string {
	return p.broker.String() + "#" + p.topic + ":" + strconv.Itoa(int(p.partitionID))
}

// SetLastConsumed sets the last consumed.
func (p *Partition) SetLastConsumed(lastConsumed bool) {
	p.lastConsumed = lastConsumed
}

// BookConsumeData sets the consumeData.
func (p *Partition) BookConsumeData(data *ConsumeData) {
	p.consumeData = data
}

// ProcConsumeResult processes consume result.
func (p *Partition) ProcConsumeResult(defHandler *flowctrl.RuleHandler, groupHandler *flowctrl.RuleHandler,
	filterConsume bool, lastConsumed bool) int64 {
	dltTime := time.Now().UnixNano()/int64(time.Millisecond) - p.consumeData.time
	p.updateStrategyData(defHandler, groupHandler)
	p.lastConsumed = lastConsumed
	switch p.consumeData.errCode {
	case errs.RetSuccess, errs.RetErrNotFound:
		if p.consumeData.msgSize == 0 && p.consumeData.errCode != errs.RetSuccess {
			p.totalZeroCnt++
		} else {
			p.totalZeroCnt = 0
		}
		if p.totalZeroCnt > 0 {
			if groupHandler.GetMinZeroCnt() != math.MaxInt64 {
				return groupHandler.GetCurFreqLimitTime(p.totalZeroCnt, p.consumeData.dltLimit) - dltTime
			} else {
				return defHandler.GetCurFreqLimitTime(p.totalZeroCnt, p.consumeData.dltLimit) - dltTime
			}
		}
		if p.consumeData.escLimit {
			return 0
		} else {
			if p.strategyData.curStageMsgSize >= p.flowCtrl.GetDataSizeLimit() ||
				p.strategyData.curSliceMsgSize >= p.strategyData.limitSliceMsgSize {
				if p.flowCtrl.GetFreqMsLimit() > p.consumeData.dltLimit {
					return p.flowCtrl.GetFreqMsLimit() - dltTime
				}
				return p.consumeData.dltLimit - dltTime
			}
			if p.consumeData.errCode == errs.RetSuccess {
				if filterConsume && p.freqCtrl.GetFreqMsLimit() >= 0 {
					if p.consumeData.requireSlow {
						return p.freqCtrl.GetZeroCnt() - dltTime
					} else {
						return p.freqCtrl.GetFreqMsLimit() - dltTime
					}
				} else if !filterConsume && p.freqCtrl.GetDataSizeLimit() >= 0 {
					return p.freqCtrl.GetDataSizeLimit() - dltTime
				}
			}
			return p.consumeData.dltLimit - dltTime
		}
	default:
		return p.consumeData.dltLimit - dltTime
	}
}

func (p *Partition) updateStrategyData(defHandler *flowctrl.RuleHandler, groupHandler *flowctrl.RuleHandler) {
	p.strategyData.curStageMsgSize += int64(p.consumeData.msgSize)
	p.strategyData.curSliceMsgSize += int64(p.consumeData.msgSize)
	curTime := time.Now().UnixNano() / int64(time.Millisecond)
	if curTime > p.strategyData.nextStageUpdate {
		p.strategyData.curStageMsgSize = 0
		p.strategyData.curSliceMsgSize = 0
		if p.consumeData.curDataDlt < 0 {
			p.flowCtrl = flowctrl.NewResult(math.MaxInt64, 20)
		} else {
			p.flowCtrl = groupHandler.GetCurDataLimit(p.consumeData.curDataDlt)
			if p.flowCtrl == nil {
				p.flowCtrl = defHandler.GetCurDataLimit(p.consumeData.curDataDlt)
			}
			if p.flowCtrl == nil {
				p.flowCtrl = flowctrl.NewResult(math.MaxInt64, 0)
			}
			p.freqCtrl = groupHandler.GetFilterCtrlItem()
			if p.freqCtrl.GetFreqMsLimit() < 0 {
				p.freqCtrl = defHandler.GetFilterCtrlItem()
			}
		}
		p.strategyData.limitSliceMsgSize = p.flowCtrl.GetDataSizeLimit() / 12
		p.strategyData.nextStageUpdate = curTime + 60000
		p.strategyData.nextSliceUpdate = curTime + 5000
	} else if curTime > p.strategyData.nextSliceUpdate {
		p.strategyData.curSliceMsgSize = 0
		p.strategyData.nextSliceUpdate = curTime + 5000
	}
}

File: dataOds/InLong-versions/inlong-release-1.3.0/inlong-tubemq/tubemq-core/src/main/java/org/apache/inlong/tubemq/corebase/cluster/Partition.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.inlong.tubemq.corebase.cluster;

import java.io.Serializable;
import org.apache.inlong.tubemq.corebase.TBaseConstants;
import org.apache.inlong.tubemq.corebase.TokenConstants;

public class Partition implements Comparable<Partition>, Serializable {

    private static final long serialVersionUID = 7587964123917870096L;

    private final BrokerInfo broker;
    private final String topic;
    private final int partitionId;
    private String partitionKey;
    private String partitionFullStr;
    private String partitionTLSFullStr;
    private int retries = 0;
    private long delayTimeStamp = TBaseConstants.META_VALUE_UNDEFINED;

    /**
     * create a Partition with broker info , topic and partitionId
     *
     * @param broker
     * @param topic
     * @param partitionId
     */
    public Partition(final BrokerInfo broker,
                     final String topic,
                     final int partitionId) {
        super();
        this.broker = broker;
        this.topic = topic;
        this.partitionId = partitionId;
        this.builderPartitionStr();
    }

    /**
     * create a Partition with full part info String
     *
     * @param strPartInfo   the partition information in string format
     */
    public Partition(String strPartInfo) {
        this.broker =
                new BrokerInfo(strPartInfo.split(TokenConstants.SEGMENT_SEP)[0]);
        String strInfo = strPartInfo.split(TokenConstants.SEGMENT_SEP)[1];
        this.topic = strInfo.split(TokenConstants.ATTR_SEP)[0];
        this.partitionId = Integer.parseInt(strInfo.split(TokenConstants.ATTR_SEP)[1]);
        this.builderPartitionStr();
    }

    /**
     * create a Partition with full broker and part info string
     *
     * @param broker
     * @param partStr
     */
    public Partition(BrokerInfo broker,
                     String partStr) {
        this.broker = broker;
        this.topic = partStr.split(TokenConstants.ATTR_SEP)[0];
        this.partitionId = Integer.parseInt(partStr.split(TokenConstants.ATTR_SEP)[1]);
        this.builderPartitionStr();
    }

    public BrokerInfo getBroker() {
        return this.broker;
    }

    public int getBrokerId() {
        return this.broker.getBrokerId();
    }

    public String getHost() {
        return this.broker.getHost();
    }

    public int getPort() {
        return this.broker.getPort();
    }

    public String getPartitionKey() {
        return this.partitionKey;
    }

    public String getTopic() {
        return topic;
    }

    public int getPartitionId() {
        return this.partitionId;
    }

    public void increRetries(int retries) {
        this.retries += retries;
        if (this.retries > 3) {
            this.delayTimeStamp = System.currentTimeMillis() + 3 * 1000;
            this.retries = 0;
        }
    }

    public void resetRetries() {
        this.retries = 0;
        this.delayTimeStamp = 0;
    }

    public int getRetries() {
        return retries;
    }

    public long getDelayTimeStamp() {
        return delayTimeStamp;
    }

    public void setDelayTimeStamp(long dltTime) {
        this.delayTimeStamp = System.currentTimeMillis() + dltTime;
    }

    @Override
    public int compareTo(final Partition other) {
        if (this.broker.getBrokerId() != other.broker.getBrokerId()) {
            return this.broker.getBrokerId() > other.broker.getBrokerId() ? 1 : -1;
        }
        if (!this.broker.getHost().equals(other.broker.getHost())) {
            return this.broker.getHost().compareTo(other.broker.getHost());
        }
        if (!this.topic.equals(other.topic)) {
            return this.topic.compareTo(other.topic);
        } else if (this.partitionId != other.partitionId) {
            return this.partitionId > other.partitionId ? 1 : -1;
        }
        return 0;
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + ((broker == null) ? 0 : broker.hashCode());
        result = prime * result + partitionId;
        result = prime * result + ((topic == null) ? 0 : topic.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        Partition other = (Partition) obj;
        if (broker == null) {
            if (other.broker != null) {
                return false;
            }
        } else if (broker.getBrokerId() != other.broker.getBrokerId()) {
            return false;
        } else if (!broker.getHost().equals(other.broker.getHost())) {
            return false;
        }
        if (partitionId != other.partitionId) {
            return false;
        }
        if (topic == null) {
            if (other.topic != null) {
                return false;
            }
        } else if (!topic.equals(other.topic)) {
            return false;
        }
        return true;
    }

    /**
     * init some string info of partition
     */
    private void builderPartitionStr() {
        StringBuilder sBuilder = new StringBuilder(256);
        this.partitionKey = sBuilder.append(this.broker.getBrokerId())
                .append(TokenConstants.ATTR_SEP).append(this.topic)
                .append(TokenConstants.ATTR_SEP).append(this.partitionId).toString();
        sBuilder.delete(0, sBuilder.length());
        this.partitionFullStr = sBuilder.append(this.broker.toString())
                .append(TokenConstants.SEGMENT_SEP).append(this.topic)
                .append(TokenConstants.ATTR_SEP).append(this.partitionId).toString();
        sBuilder.delete(0, sBuilder.length());
        this.partitionTLSFullStr = sBuilder.append(this.broker.getFullTLSInfo())
                .append(TokenConstants.SEGMENT_SEP).append(this.topic)
                .append(TokenConstants.ATTR_SEP).append(this.partitionId).toString();

    }

    @Override
    public String toString() {
        return this.partitionFullStr;
    }

    /**
     * get the partition full string info
     *
     * @param overTLS
     * @return
     */
    public String getPartitionFullStr(boolean overTLS) {
        if (overTLS) {
            return this.partitionTLSFullStr;
        } else {
            return this.partitionFullStr;
        }

    }

}
